Fixed bugs to ensure proper flow of prompts and remove repetitive prompts

Aad1tya27 opened the pull request:
**What kind of change does this PR introduce?**
Bug in setup.ts

**Issue Number:**

Fixes #2829  <!--Add related issue number here.-->

**Snapshots/Videos:**
![image](https://github.com/user-attachments/assets/f1ba79b4-4865-427a-9225-e35ef327422e)

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**
No
<!--Add link to Talawa-Docs.-->

**Does this PR introduce a breaking change?**
Yeah, there was a breaking change related to __dirname, which was caused due CommonJS to ES.
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

## Checklist

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [ ] I have written tests for all new changes/features
- [ ] I have verified that test coverage meets or exceeds 95%
- [ ] I have run the test suite locally and all tests pass


**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

<!--Yes or No-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**
	- Enhanced error handling and clarity in the setup process.
	- Improved data importation logic with new functions for Docker and non-Docker scenarios.
	- Updated path resolution mechanisms for transaction logs.
	- Refined database connection logic with retry capabilities.
- **Tests**
	- Introduced a new test suite for validating the transaction log path input functionality.
	- Added comprehensive tests for data importation with and without Docker, covering various user scenarios and error conditions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

palisadoes reviewed (changes requested):
1. Remember to add tests. 
2. We can't have this again

Aad1tya27 commented:
> 1. Remember to add tests.
> 2. We can't have this again

Yes I'm writing tests now.

Aad1tya27 commented on setup.ts:
```diff
@@ -912,6 +915,126 @@ export async function configureMinio(
+      await importDefaultData();
+    }
+  }
+}
```
I've added this suggestion in the latest commit

Aad1tya27 commented on tests/setup/dataImportFlow.spec.ts:
```diff
@@ -0,0 +1,399 @@
+    expect(importDefaultDataMock).not.toBeCalled();
+    expect(importDataMock).not.toBeCalled();
+  });
+});
```
I've added this test case in the latest commit

Aad1tya27 commented on tests/setup/dataImportFlow.spec.ts:
```diff
@@ -0,0 +1,437 @@
+    expect(importDataMock).not.toBeCalled();
+  });
+
+  it("should terminate execution if error is encountered while starting containers", async () => {
+    const runDockerComposeMock = vi
+      .fn()
+      .mockImplementation(async (): Promise<void> => {
+        throw new Error("Error starting containers");
+      });
+    const importDataMock = vi
+      .fn()
+      .mockImplementation(async (): Promise<void> => {
+        return Promise.resolve();
+      });
+    const importDefaultDataMock = vi
+      .fn()
+      .mockImplementation(async (): Promise<void> => {
+        return Promise.resolve();
+      });
+    const connectDatabaseMock = vi
+      .fn()
+      .mockImplementation(async (): Promise<void> => {
+        return Promise.resolve();
+      });
+    vi.spyOn(inquirer, "prompt").mockResolvedValueOnce({
+      shouldStartDockerContainers: true,
+    });
+
+    await dataImportWithDocker(
+      runDockerComposeMock,
+      importDefaultDataMock,
+      importDataMock,
+      connectDatabaseMock,
+    );
+    expect(runDockerComposeMock).toBeCalled();
+    expect(connectDatabaseMock).not.toBeCalled();
+    expect(importDefaultDataMock).not.toBeCalled();
+    expect(importDataMock).not.toBeCalled();
+  });
```
I've added this test in the latest commit.

Aad1tya27 commented on tests/setup/dataImportFlow.spec.ts:
```diff
@@ -0,0 +1,464 @@
+        return Promise.resolve();
+      });
+    vi.spyOn(inquirer, "prompt")
+      .mockResolvedValueOnce({ shouldOverwriteData: true })
+      .mockResolvedValueOnce({ importSampleData: true });
```
In the setup.ts file they are using importSampleData in one prompt and shouldImportSampleData in the other, and I've added them accordingly. This won't cause any tests to fail.

Aad1tya27 commented on tests/setup/dataImportFlow.spec.ts:
```diff
@@ -0,0 +1,464 @@
+    expect(wipeExistingDataMock).not.toBeCalled();
+    expect(importDefaultDataMock).toBeCalled();
+    expect(importDataMock).not.toBeCalled();
+  });
```
I've corrected the test name in the latest commit.

palisadoes commented:
Your screen shots don't show the user being offered the opportunity to install sample data after saying No to the default data.

Aad1tya27 commented:
> Your screen shots don't show the user being offered the opportunity to install sample data after saying No to the default data.

There is no option to say No to the default data(which is just the organizations). There is only option to reject Sample Data, and on doing do default data will be automatically imported.

Aad1tya27 commented:
If Docker is not Used:
Case 1: No Collections Found, Rejected Sample Data => Default Data is loaded
![image](https://github.com/user-attachments/assets/ff3d885f-3d14-4e4e-b072-4f849708ce9e)


Case 2: No Collections Found, Accepted Sample Data => Sample Data is loaded
![image](https://github.com/user-attachments/assets/d77959c2-35ad-44b9-b196-7199d0629eeb)
![image](https://github.com/user-attachments/assets/3c6a1cfd-8afe-44d1-8c85-cc83d11d3ba6)
(Here there is no repetition anymore.)


Case 3: Collections Found, Data not overwritten => No change
![image](https://github.com/user-attachments/assets/dae2a7c2-8d98-4b1d-8a6d-31f3e4363b30)


Case 4: Collections Found, Data overwritten => option between loading sample data or default data
![image](https://github.com/user-attachments/assets/5464b55c-8de9-4b65-9cdf-cae06a455337)

Aad1tya27 commented:
@palisadoes If you want to see screenshots of Docker installation I can share them too, and if any other change is required let me know.

palisadoes commented:
The logic must be:

1. If data **_IS_** found, then you must be prompted to import it.
    1. If you select `Yes` :
        1. You are prompted to import the default data.
            1. If you import the default data, then there must be no prompt for importing the sample data
        1. If you do not import the default data, you are prompted to import the default data.
            2. If you do not choose to import the default data, then it's a no-op
1. If data **_IS NOT_** found, then you must be prompted to import it. 
    1. If you select `Yes` 
        1. You are prompted to import the default data.
            1. If you import the default data, then there must be no prompt for importing the sample data
        1. If you do not import the default data, you are prompted to import the default data.
            2. If you do not choose to import the default data, then it's a no-op
   1. If you select `No`, then it's a no-op 

Other comments:

The questions are too wordy.

Instead of the prompt : 
    - `Do you want to delete the existing data and import required default Data to start using Talawa?` 
have:
    - `Do you want to DESTROY the existing data, replacing it with the default data required for a fresh production system?`

Instead of the prompt : 
    - `Do you want to import Talawa sample data for testing and evaluation purposes?` 
have:
    - `Do you want to DESTROY the existing data, replacing it with data suitable for testing and evaluation?`

Aad1tya27 commented:
@palisadoes I've corrected the flow of prompts and changed the tests accordingly. Kindly re-review it. For some reason I'm failing the Validate Code Rabbit check, even though it approved my changes.
